### PR TITLE
Reject trailing bytes

### LINKED ISSUE #193: Reject trailing bytes when deserializing from &[u8]
In `deserialize_from`, it makes sense to allow deserializing multiple values from the same io::Read. But in `deserialize`, I think trailing data should be an error.

I would expect this to be an error rather than ignoring `b"x"`:

```rust
extern crate bincode;

fn main() {
    println!("{:?}", bincode::deserialize::<char>(b"1x"));
}
```

### PR DESCRIPTION
**Pull request by ZoeyR**
Closes #193 

There were two design possibilities I considered for tackling this problem. The current one changes the impls on `SliceReader` to be impls for `&mut SliceReader`. This allows me to access it after passing it to the deserializer. The other option would be to allow `pub(restricted)` access to the reader field of the deserializer and access it from there.

### CODE REVIEW

**Comment by ZoeyR** [comment]
Finally got around to updating this to current bincode after 2.5 years 😅

File: src/de/read.rs [hunk 1]
```diff
@@ -101,7 +105,7 @@ impl<'storage> SliceReader<'storage> {
     }
 }
 
-impl<'storage> BincodeRead<'storage> for SliceReader<'storage> {
+impl<'a, 'storage> BincodeRead<'storage> for &'a mut SliceReader<'storage> {
```
Comment (jdm) [inline src/de/read.rs]: Why are these changes necessary?

Comment (ZoeyR) [reply src/de/read.rs]: I guess now that we use `pub(crate)` throughout the coebase I can use the other option I outlined in my PR. Guess that happens when you don't update a PR for two years lol

**Comment by ZoeyR** [comment]
I've realized that this might be considered a breaking change.

**Comment by strega-nil** [comment]
I was the person who was saying that this should be considered a breaking change, fwiw.

**Comment by ZoeyR** [comment]
cc @strega-nil
